#' plot the corresponding binary heatmap| ccf heatmaf of the NJtree
#'
#' @param maf an \code{\link{maf}} object generated by \code{\link{read.maf}
#' @param is.ccf Use mutation CCF for heatmap painting. Default FALSE.
#' @param use.indel logical. Whether to uses all variants irrespective of values in Mutation_Status.Defaults to FALSE, only use somatic mutations.
#' @param use.ccf logical. Whether to print the binary heatmap with ccf values. Defaults to FALSE, print the binary heatmap without ccf value
#' @param ccf
#' @param ccf.mutation.id manually specify which columns could be joint by ccf.mutation.sep to get the same format of mutation id in ccfy
#' @param ccf.mutation.sep manually specify the separator character.Values on each line of the ccf.mutation.id are separated by this character. I
#' @return binary mutation heatmap or mutation CCF heatmap
#' @export
#'
#' @examples
#' mut.heatmap(maf, use.indel = F)
#' mut.heatmap(maf, use.indel = F, use.ccf = TRUE)

heatmap_input <- function(maf, use.indel, use.ccf, ccf, ccf.mutation.id, ccf.mutation.sep){
  if(use.ccf){
    names <- c("mutation", "sample", "CCF")
    mut_sort <- mut_ccf_sort(snp, ccf = ccf, use.indel = use.indel, ccf.mutation.id, ccf.mutation.sep)

  }
  else{
    names <- c("mutation", "sample", "Mutation")
    mut_sort <- mut_binary_sort(maf, use.indel = use.indel)
  }

  mut_dat <- melt(mut_dat)
  colnames(mut_dat) <- names
  return(mut_dat)
}



mut.heatmap <- function(maf, use.indel = FALSE, use.ccf = FALSE, ccf = data.frame(), ccf.mutation.id = c("Hugo_Symbol","Chromosome","Start_Position"), ccf.mutation.sep = ":"){
  type  <- "Mutation"
  if(use.ccf){
    type <- "CCF"
  }

  mut_dat <- heatmap_input(maf, use.indel = use.indel, use.ccf = use.ccf, ccf = ccf, ccf.mutation.id = ccf.mutation.id, ccf.mutation.sep = ccf.mutation.sep)
  p_basic <- ggplot(mut_dat, aes(sample, mutation)) +
    labs(x = "", y = "") + theme_bw() +
    theme(panel.border = element_blank()) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
    theme(axis.text.y = element_blank()) +
    theme(axis.text.x = element_text(angle = 45, vjust = 1,size = 10, hjust = 1, face = "bold")) +
    theme(axis.ticks = element_blank()) +
    theme(legend.title = element_text(size = 12, face = "bold", color = "black")) +
    theme(legend.text = element_text(size = 10, face = "bold", color = "black")) +
    labs(fill = type) + scale_y_continuous(expand = c(0,0))

  if(use.ccf){
    p <- p_basic + geom_tile(aes(fill = mut_sort$CCF)) + scale_fill_gradient(low = "grey", high = "red", na.value="black", limit=c(0, 1))
    ggsave(paste(patientID, "_mut_CCF.pdf", sep = ""), p, width = 4.5, height = 6.5)
    }
  if(!use.ccf){
    p <- p_basic + geom_tile(aes(fill = as.factor(mut_sort$Mutation))) + scale_fill_manual(values = c("grey", "red"))
    ggsave(paste(patientID, "_mut.pdf", sep = ""), p, width = 4.5, height = 6.5)
    }
}


